src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java

```diff
@@ -419,8 +417,30 @@ private JlinkConfiguration initJlinkConfig() throws BadArgs {
         Set<String> roots = new HashSet<>();
         for (String mod : options.addMods) {
             if (mod.equals(ALL_MODULE_PATH)) {
-                ModuleFinder mf = newLimitedFinder(finder, options.limitMods,
-                                              Set.of());
+                // all observable modules in the app module path are roots
+                Set<String> initialRoots = appModuleFinder.findAll()
+                        .stream()
+                        .map(ModuleReference::descriptor)
+                        .map(ModuleDescriptor::name)
+                        .collect(Collectors.toSet());
+
+                // Error if no module is found on the app module path
+                if (initialRoots.isEmpty()) {
+                    String modPath = originalModulePath.stream()
+                            .map(a -> a.toString())
+                            .collect(Collectors.joining(", "));
+                    throw taskHelper.newBadArgs("err.empty.module.path", modPath);
+                }
+
+                // Use a module finder with limited observability, as determined
+                // by initialRoots, to find the observable modules from the
+                // application module path (--module-path option) only. We must
+                // not include JDK modules from the default module path or the
+                // run-time image. Only do this if no --limit-modules has been
+                // specified to begin with.
+                ModuleFinder mf = newLimitedFinder(finder,
```

Comment:
FWIW, rereading this, it's still a bit awk to call a module finder a limited finder.

Reply:
How about `newScopedFinder()`?

Reply:
Best to leave out renaming when in doubt - either `newModuleFinder` or `limitFinder`.

Reply:
OK. I'll go with `limitFinder`.